### PR TITLE
Experimental, parallel test execution

### PR DESCRIPTION
**Pull request by stegut01**
# Description

Added new experimental feature: thread model for running mulitple test jobs in parallel on multiple resources instances (boards) of the same type within greentea, where for each board we have a separate thread.

Currently each unique MUT (mbed under test) is flashed separately one by one, but with this new feature we are able to flash multiple boards at the same time, run test case on hardware and get results in parallel.

Test execution time is reduced by n where n is number of MUTs (MUT: mbed under test).

## Limitations

At the moment the parallelisation is only for resource instances of the **same type**.
Don't use the ```--verbose-test-result``` mode with the new ```--parallel``` mode.

# Usage

To test in parallel use:
```
$ mbedgt -t frdm-k64f-gcc --parallel 2
```
Where ```-t``` specifies the target and ```--parallel``` enables parallel testing with an *int* argument, which specifies the number of boards used in parallel. Logically, it is not possible to test on more devices as connected.

## Example

We have three K64F platforms connected:
```
$ mbedls
+---------------------+----------------------------+-------------------+-------------------+--------------------------------+
|platform_name        |platform_name_unique        |mount_point        |serial_port        |target_id                       |
+---------------------+----------------------------+-------------------+-------------------+--------------------------------+
|K64F                 |K64F[0]                     |E:                 |COM12              |02400201A0B31E095D4DE3B1        |
|K64F                 |K64F[1]                     |G:                 |COM14              |02400201AA3F4E7757C1B3CF        |
|K64F                 |K64F[2]                     |H:                 |COM15              |02400201AA244E7657DAB3CE        |
+---------------------+----------------------------+-------------------+-------------------+--------------------------------+
```
We want to force Greentea to run the tests on 2 boards and divide the test cases, so that we can achieve a speedup of about 2x.

We will show example with tests from mbed-drivers repository. 
```
$ cd mbed-drivers
$ mbedgt -t frdm-k64f-gcc --parallel 2
```
```
mbedgt: detecting connected mbed-enabled devices...
mbedgt: detected 3 devices
        detected 'K64F' -> 'K64F[0]', console at 'COM12', mounted at 'E:', target id '02400201A0B31E095D4DE3B1'
        detected 'K64F' -> 'K64F[1]', console at 'COM14', mounted at 'G:', target id '02400201AA3F4E7757C1B3CF'
        detected 'K64F' -> 'K64F[2]', console at 'COM15', mounted at 'H:', target id '02400201AA244E7657DAB3CE'
mbedgt: yotta search for mbed-target 'k64f'
        calling yotta: yotta --plain search -k mbed-target:k64f target
        found target 'frdm-k64f-gcc'
        found target 'frdm-k64f-armcc'
mbedgt: processing 'frdm-k64f-gcc' yotta target compatible platforms...
mbedgt: processing 'K64F' platform...
mbedgt: using platform 'K64F' for test:
        target_id_mbed_htm = '02400201A0B31E095D4DE3B1'
        mount_point = 'E:'
        target_id = '02400201A0B31E095D4DE3B1'
        serial_port = 'COM12'
        target_id_usb_id = '02400201A0B31E095D4DE3B1'
        platform_name = 'K64F'
        platform_name_unique = 'K64F[0]'
mbedgt: using platform 'K64F' for test:
        target_id_mbed_htm = '02400201AA3F4E7757C1B3CF'
        mount_point = 'G:'
        target_id = '02400201AA3F4E7757C1B3CF'
        serial_port = 'COM14'
        target_id_usb_id = '02400201AA3F4E7757C1B3CF'
        platform_name = 'K64F'
        platform_name_unique = 'K64F[1]'
mbedgt: building your sources and tests with yotta...
        calling yotta: yotta --target=frdm-k64f-gcc,* build
info: generate for target: frdm-k64f-gcc 0.1.3 at C:\Users\stefan.gutmann@arm.com\Documents\GitHub\mbed-drivers\yotta_targets\frdm-k64f-gcc
GCC version is: 4.9.3
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/stefan.gutmann@arm.com/Documents/GitHub/mbed-drivers/build/frdm-k64f-gcc
ninja: no work to do.
mbedgt: yotta build for target 'frdm-k64f-gcc' was successful
mbedgt: running 23 tests for target 'frdm-k64f-gcc' and platform 'K64F'
        use 2 instances for parallel testing
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-echo' .......................................................... OK in 6.32 sec
mbedgt: test on hardware with target id: 02400201AA3F4E7757C1B3CF
        test 'mbed-drivers-test-time_us' ....................................................... OK in 11.37 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-serial_interrupt' .............................................. OK in 6.17 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-functionpointer' ............................................... OK in 1.38 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-stdio' ......................................................... OK in 0.74 sec
mbedgt: test on hardware with target id: 02400201AA3F4E7757C1B3CF
        test 'mbed-drivers-test-blinky' ........................................................ TIMEOUT in 20.93 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-eventhandler' .................................................. ERROR in 0.48 sec
mbedgt: test on hardware with target id: 02400201AA3F4E7757C1B3CF
        test 'mbed-drivers-test-stl' ........................................................... OK in 1.37 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-div' ........................................................... OK in 1.38 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-cstring' ....................................................... FAIL in 1.39 sec
mbedgt: test on hardware with target id: 02400201AA3F4E7757C1B3CF
        test 'mbed-drivers-test-rtc' ........................................................... OK in 4.58 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-cpp' ........................................................... OK in 1.39 sec
mbedgt: test on hardware with target id: 02400201AA3F4E7757C1B3CF
        test 'mbed-drivers-test-timeout' ....................................................... OK in 11.52 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-ticker_3' ...................................................... OK in 11.38 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-heap_and_stack' ................................................ OK in 1.38 sec
mbedgt: test on hardware with target id: 02400201AA3F4E7757C1B3CF
        test 'mbed-drivers-test-ticker_2' ...................................................... OK in 11.35 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-hello' ......................................................... OK in 0.38 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-dev_null' ...................................................... OK in 3.47 sec
mbedgt: test on hardware with target id: 02400201AA3F4E7757C1B3CF
        test 'mbed-drivers-test-ticker' ........................................................ OK in 11.37 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-basic' ......................................................... OK in 1.36 sec
mbedgt: test on hardware with target id: 02400201AA3F4E7757C1B3CF
        test 'mbed-drivers-test-asynch_spi' .................................................... ERROR in 0.41 sec
mbedgt: test on hardware with target id: 02400201A0B31E095D4DE3B1
        test 'mbed-drivers-test-sleep_timeout' ................................................. OK in 3.36 sec
mbedgt: test on hardware with target id: 02400201AA3F4E7757C1B3CF
        test 'mbed-drivers-test-detect' ........................................................ FAIL in 0.42 sec
mbedgt: test report:
+---------------+---------------+------------------------------------+---------+--------------------+-------------+
| target        | platform_name | test                               | result  | elapsed_time (sec) | copy_method |
+---------------+---------------+------------------------------------+---------+--------------------+-------------+
| frdm-k64f-gcc | K64F          | mbed-drivers-test-div              | OK      | 1.38               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-rtc              | OK      | 4.58               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-cstring          | FAIL    | 1.39               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-cpp              | OK      | 1.39               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-timeout          | OK      | 11.52              | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-ticker_3         | OK      | 11.38              | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-ticker_2         | OK      | 11.35              | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-heap_and_stack   | OK      | 1.38               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-hello            | OK      | 0.38               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-time_us          | OK      | 11.37              | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-ticker           | OK      | 11.37              | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-echo             | OK      | 6.32               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-serial_interrupt | OK      | 6.17               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-blinky           | TIMEOUT | 20.93              | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-functionpointer  | OK      | 1.38               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-stdio            | OK      | 0.74               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-stl              | OK      | 1.37               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-dev_null         | OK      | 3.47               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-basic            | OK      | 1.36               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-eventhandler     | ERROR   | 0.48               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-asynch_spi       | ERROR   | 0.41               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-sleep_timeout    | OK      | 3.36               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-detect           | FAIL    | 0.42               | shell       |
+---------------+---------------+------------------------------------+---------+--------------------+-------------+

Result: 2 FAIL / 18 OK / 1 TIMEOUT / 2 ERROR
completed in 138.94 sec
```

### CODE REVIEW

**Comment by PrzemekWirkus** [comment]
Comments:
* Please remove CammelCase from prints (only use lowercase text) - we want to allign with how yotta prints stuff
* Please make sure Ctrl+C stops parallel execution of Greentea

**Comment by PrzemekWirkus** [comment]
You are calling yotta (```yotta --target=frdm-k64f-gcc,* build```) twice for the same thing:
```
mbedgt: building your sources and tests with yotta...
        calling yotta: yotta --target=frdm-k64f-gcc,* build
info: generate for target: frdm-k64f-gcc 0.1.4 at c:\Work\mbed-drivers\yotta_targets\frdm-k64f-gcc
GCC version is: 4.8.4
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Work/mbed-drivers/build/frdm-k64f-gcc
ninja: no work to do.
mbedgt: yotta build for target 'frdm-k64f-gcc' was successful
mbedgt: running 20 tests for target 'frdm-k64f-gcc' and platform 'K64F'
mbedgt: building your sources and tests with yotta...
        calling yotta: yotta --target=frdm-k64f-gcc,* build
info: generate for target: frdm-k64f-gcc 0.1.4 at c:\Work\mbed-drivers\yotta_targets\frdm-k64f-gcc
GCC version is: 4.8.4
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Work/mbed-drivers/build/frdm-k64f-gcc
```

**Comment by PrzemekWirkus** [comment]
When used with ```-n``` and odd number of test error occures:
```
$ mbedgt --parallel 3 -n mbed-drivers-test-hello
```
```
mbedgt: checking for yotta target in current directory
        reason: no --target switch set
mbedgt: checking yotta target in current directory
        calling yotta: yotta --plain target
mbedgt: assuming default target as 'frdm-k64f-gcc'
mbedgt: detecting connected mbed-enabled devices...
mbedgt: detected 3 devices
        detected 'K64F' -> 'K64F[0]', console at 'COM160', mounted at 'E:', target id '024002031E031E6AE3FFE3D2'
        detected 'K64F' -> 'K64F[1]', console at 'COM162', mounted at 'F:', target id '02400203C3423E603EBEC3D8'
        detected 'NUCLEO_F401RE' -> 'NUCLEO_F401RE[0]', console at 'COM163', mounted at 'G:', target id '07200200076165023804F31F'
mbedgt: yotta search for mbed-target 'k64f'
        calling yotta: yotta --plain search -k mbed-target:k64f target
        found target 'frdm-k64f-gcc'
        found target 'frdm-k64f-armcc'
...
...
mbedgt: yotta build for target 'frdm-k64f-gcc' was successful
mbedgt: test case filter (specified with -n option)
        test filtered in 'mbed-drivers-test-hello'
        use 2 instances for parallel testing
mbedgt: Unexpected error:
        list index out of range
Traceback (most recent call last):
  File "C:\Python27\Scripts\mbedgt-script.py", line 9, in <module>
    load_entry_point('mbed-greentea==0.1.7', 'console_scripts', 'mbedgt')()
  File "c:\work\stefan\greentea\mbed_greentea\mbed_greentea_cli.py", line 224, in main
    cli_ret = main_cli(opts, args)
  File "c:\work\stefan\greentea\mbed_greentea\mbed_greentea_cli.py", line 598, in main_cli
    if temp_test_report.keys()[0] not in test_report:
IndexError: list index out of range
```

**Comment by jupe** [comment]
I didn't saw any thread joins...

File: mbed_greentea/mbed_greentea_cli.py [hunk 1]
```diff
@@ -371,14 +442,26 @@ def main_cli(opts, args, gt_instance_uuid=None):
...
     test_platforms_match = 0    # Count how many tests were actually ran with current settings
     target_platforms_match = 0  # Count how many platforms were actually tested with current settings
 
-    test_report = {}    # Test report used to export to Junit, HTML etc...
-
+    test_report = {}            # Test report used to export to Junit, HTML etc...
+    muts_to_test = []           # MUTs to actually be tested
+    test_queue = Queue()        # contains information about test_bin and image_path for each test case
+    test_result_queue = Queue() # used to store results of each thread
+    execute_threads = []        # list of threads to run test cases
```
Comment (jupe) [inline mbed_greentea/mbed_greentea_cli.py]: there is available threadpool, which could be better choose than array of threads

File: mbed_greentea/mbed_greentea_cli.py [hunk 2]
```diff
@@ -371,14 +442,26 @@ def main_cli(opts, args, gt_instance_uuid=None):
...
-
     test_exec_retcode = 0       # Decrement this value each time test case result is not 'OK'
     test_platforms_match = 0    # Count how many tests were actually ran with current settings
     target_platforms_match = 0  # Count how many platforms were actually tested with current settings
 
-    test_report = {}    # Test report used to export to Junit, HTML etc...
-
+    test_report = {}            # Test report used to export to Junit, HTML etc...
+    muts_to_test = []           # MUTs to actually be tested
+    test_queue = Queue()        # contains information about test_bin and image_path for each test case
```
Comment (jupe) [inline mbed_greentea/mbed_greentea_cli.py]: test_bin could be parameter for test thread. then this Queue is not needed.

**Pull request merged**